pkg/cdi/cache.go

```diff
@@ -277,3 +339,113 @@ func (c *Cache) GetSpecDirectories() []string {
 	copy(dirs, c.specDirs)
 	return dirs
 }
+
+// Our fsnotify helper wrapper.
+type watch struct {
+	w       *fsnotify.Watcher
+	watched map[string]bool
+}
+
+// Start watching a set of directories for relevant changes.
+func (w *watch) start(dirs []string, m *sync.Mutex, refresh func() error) error {
+	var (
+		dir string
+		err error
+	)
+
+	w.watched = make(map[string]bool)
+
+	w.w, err = fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Wrap(err, "failed to create Spec dir watch")
+	}
+
+	for _, dir = range dirs {
+		w.watched[dir] = false
+	}
+
+	w.update()
+	go w.watch(m, refresh)
+	refresh()
+
+	return nil
+}
+
+// Stop watching directories.
+func (w *watch) stop() {
+	if w.w == nil {
+		return
+	}
+
+	w.w.Close()
+	w.watched = nil
+}
+
+// Update watch with pending/missing or removed directories.
+func (w *watch) update(removed ...string) bool {
+	var (
+		dir    string
+		done   bool
+		err    error
+		update bool
+	)
+
+	for dir, done = range w.watched {
+		if done {
+			continue
+		}
+		err = w.w.Add(dir)
+		switch {
+		case err == nil:
+			w.watched[dir] = true
+			update = true
+		case !os.IsNotExist(err):
+			fallthrough
+		default:
+			w.watched[dir] = false
+		}
+	}
+
+	for _, dir = range removed {
```

Comment:
Does it not make sense to remove these first and then update `w.watched`?

Reply:
It can be done like that, too. I'm not sure though if that matters at all. I think it does not. The idea here is that we always check here if we could add a pending spec directory, because we don't get any external events for the creation of those. We do get events from subsequent removal, which we then handle, if we were triggered by a removal event.